subprojects/configuration-cache/src/integTest/groovy/org/gradle/configurationcache/ConfigurationCacheValueSourceIntegrationTest.groovy

```diff
@@ -367,4 +368,160 @@ class ConfigurationCacheValueSourceIntegrationTest extends AbstractConfiguration
         configurationCache.assertStateStored()
         outputContains("ValueSource result = Hello, world")
     }
+
+    def "value source can read mutated system property inputs at configuration time"() {
+        given:
+        def configurationCache = newConfigurationCacheFixture()
+        buildFile("""
+        import org.gradle.api.provider.*
+
+        abstract class IdentitySource implements ValueSource<String, Parameters> {
+            interface Parameters extends ValueSourceParameters {
+                Property<String> getValue()
+            }
+
+            @Override String obtain() {
+                return parameters.value.orNull
+            }
+        }
+
+        def vsResult = providers.of(IdentitySource) {
+            parameters.value = providers.systemProperty("property")
+        }
+
+        System.setProperty("property", "someValue")
+
+        println("configuration value = \${vsResult.getOrElse("NO VALUE")}")
+
+        tasks.register("echo") {
+            doLast {
+                println("execution value = \${vsResult.getOrElse("NO VALUE")}")
+            }
+        }
+
+        defaultTasks "echo"
+        """)
+
+        when:
+        configurationCacheRun()
+
+        then:
+        configurationCache.assertStateStored()
+        outputContains("configuration value = someValue")
+        outputContains("execution value = someValue")
+
+        when:
+        configurationCacheRun()
+
+        then:
+        configurationCache.assertStateLoaded()
+        outputContains("execution value = someValue")
+    }
+
+    @ToBeImplemented("https://github.com/gradle/gradle/issues/23689")
+    def "value source can read mutated system property at configuration time with Java API"() {
```

Comment:
🧪 Perhaps a similar test for environment variables?

Reply:
>Perhaps a similar test for environment variables?

There's no API to alter environment variables at runtime without using unforgivable curses, so I don't think the test for it is valuable enough.

Reply:
Ah, right.